### PR TITLE
periodic boundary condition and homogenization

### PR DESCRIPTION
**Pull request by Huangzizhou**

### CODE REVIEW

**Comment by Huangzizhou** [comment]
TODO:
- [ ] Periodic linear problem: Linear Elasticity, Poisson
- [ ] Periodic in single direction
- [ ] Unit test for 3D
- [ ] Differentiability

File: src/polyfem/State.cpp [hunk 1]
```diff
@@ -882,6 +911,8 @@ namespace polyfem
 
 		logger().info("Building collision mesh...");
 		build_collision_mesh();
+		if (periodic_bc && args["contact"]["periodic"])
```
Comment (teseoch) [inline src/polyfem/State.cpp:914]: do you need periodic contact?

File: src/polyfem/State.cpp [hunk 2]
```diff
@@ -966,6 +997,36 @@ namespace polyfem
 		auto it = std::unique(boundary_nodes.begin(), boundary_nodes.end());
 		boundary_nodes.resize(std::distance(boundary_nodes.begin(), it));
 
+		// for elastic pure periodic problem, find an internal node and force zero dirichlet
+		if ((!problem->is_time_dependent() || args["time"]["quasistatic"]) && boundary_nodes.size() == 0 && has_periodic_bc())
```
Comment (teseoch) [inline src/polyfem/State.cpp:1001]: is this the standard way?

File: src/polyfem/State.cpp [hunk 3]
```diff
@@ -1030,9 +1091,9 @@ namespace polyfem
 
 		out_geom.build_grid(*mesh, args["output"]["advanced"]["sol_on_grid"]);
 
-		if (!problem->is_time_dependent() && boundary_nodes.empty())
+		if ((!problem->is_time_dependent() || args["time"]["quasistatic"]) && boundary_nodes.empty())
 		{
-			log_and_throw_error("Static problem need to have some Dirichlet nodes!");
+			logger().warn("(Quasi-)Static problem without Dirichlet nodes, will fix solution at one node to find a unique solution!");
```
Comment (teseoch) [inline src/polyfem/State.cpp:1096]: why we have quasi static, what is the difference?

File: src/polyfem/assembler/MacroStrain.hpp [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+                {
+                    json arg = param["linear_displacement_offset"];
+                    assert(arg.is_array() && arg.size() > 0 && arg[0].is_array());
+                    
+                    for (size_t i = 0; i < arg.size(); ++i)
+                    {
+                        for (size_t j = 0; j < arg[i].size(); ++j)
+                        {
+                            value[i * 3 + j].init(arg[i][j]);
+                            value[i * 3 + j].set_unit_type("");
```
Comment (teseoch) [inline src/polyfem/assembler/MacroStrain.hpp:40]: shudnt be displacement?

File: src/polyfem/assembler/MacroStrain.hpp [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+                for (int i = 0; i < _dim; i++)
+                    for (int j = 0; j < _dim; j++)
+                        strain(i, j) = eval(i, j, t);
+                
+                return strain;
+            }
+
+			double eval(const int i, const int j, const double t) const
+            {
+                return value[i * 3 + j](0,0,0,t);
```
Comment (teseoch) [inline src/polyfem/assembler/MacroStrain.hpp:58]: 0,0,0?

File: src/polyfem/assembler/MacroStrain.hpp [hunk 6]
```diff
@@ -0,0 +1,62 @@
+#pragma once
```
Comment (teseoch) [inline src/polyfem/assembler/MacroStrain.hpp:1]: why is this inside assembler?

File: src/polyfem/state/StateInit.cpp [hunk 7]
```diff
@@ -252,6 +252,7 @@ namespace polyfem
 		{
 			args["solver"]["contact"]["friction_iterations"] = 0;
 			args["contact"]["friction_coefficient"] = 0;
+			args["contact"]["periodic"] = false;
```
Comment (teseoch) [inline src/polyfem/state/StateInit.cpp:255]: maybe remove?

**Review by teseoch — COMMENTED** [review]
Maybe add some integration tests?

File: src/polyfem/State.cpp (hunk 2 above)
Comment (Huangzizhou) [reply src/polyfem/State.cpp:1001]: This is one of the standard ways, another way is to enforce zero mean as a Lagrangian multiplier, which requires solving a saddle point problem.

File: src/polyfem/assembler/MacroStrain.hpp (hunk 4 above)
Comment (Huangzizhou) [reply src/polyfem/assembler/MacroStrain.hpp:40]: It's a 3x3 matrix, independent of spatial coordinate, but can vary over time. It has the same unit as deformation gradient, which is unit-less.

File: src/polyfem/assembler/MacroStrain.hpp (hunk 5 above)
Comment (Huangzizhou) [reply src/polyfem/assembler/MacroStrain.hpp:58]: It's a 3x3 matrix, independent of spatial coordinate, but can vary over time.

File: src/polyfem/assembler/MacroStrain.hpp (hunk 6 above)
Comment (Huangzizhou) [reply src/polyfem/assembler/MacroStrain.hpp:1]: I consider this as some kind of "boundary condition", so I place it in the same folder as RHSAssembler

**Comment by Huangzizhou** [comment]
> Maybe add some integration tests?

The tests were added to the test_runners list, both 2D and 3D periodic solve.

File: src/polyfem/State.cpp (hunk 3 above)
Comment (Huangzizhou) [reply src/polyfem/State.cpp:1096]: Quasi-static has been there for a long time, it's transient without inertia, but I guess no one was using it. I added quasi-static here because the linear system is also singular just as static.
